chain/client/src/chunk_validation.rs

```diff
@@ -0,0 +1,153 @@
+            self.epoch_manager.get_block_producer(&epoch_id, chunk_header.height_created())?;
+
+        let network_sender = self.network_sender.clone();
+        let signer = self.my_signer.clone().unwrap();
+        let runtime_adapter = self.runtime_adapter.clone();
+        rayon::spawn(move || match validate_chunk(&state_witness, &*runtime_adapter) {
+            Ok(()) => {
+                tracing::debug!(
+                    target: "chunk_validation",
+                    "Chunk {:?} validated successfully, sending approval to {:?}",
+                    chunk_header.chunk_hash(),
+                    block_producer,
+                );
+                let endorsement_to_sign = ChunkEndorsementInner::new(chunk_header.chunk_hash());
+                network_sender.send(PeerManagerMessageRequest::NetworkRequests(
+                    NetworkRequests::ChunkEndorsement(ChunkEndorsementMessage {
+                        approval: ChunkEndorsement {
+                            account_id: signer.validator_id().clone(),
+                            signature: signer.sign_chunk_endorsement(&endorsement_to_sign),
+                            inner: endorsement_to_sign,
+                        },
+                        target: block_producer,
+                    }),
+                ));
+            }
+            Err(err) => {
+                tracing::error!("Failed to validate chunk: {:?}", err);
+            }
+        });
+        Ok(())
+    }
+}
+
+/// The actual chunk validation logic.
+fn validate_chunk(
+    state_witness: &ChunkStateWitness,
+    runtime_adapter: &dyn RuntimeAdapter,
+) -> Result<(), Error> {
+    // TODO: Replace this with actual stateless validation logic.
+    if state_witness.state_root != state_witness.chunk_header.prev_state_root() {
+        return Err(Error::InvalidChunkStateWitness);
+    }
+    // We'll need the runtime no matter what so just leaving it here to avoid an
+    // unused variable warning.
+    runtime_adapter.get_tries();
+    Ok(())
+}
+
+impl Client {
+    /// Responds to a network request to verify a `ChunkStateWitness`, which is
+    /// sent by chunk producers after they produce a chunk.
+    pub fn process_chunk_state_witness(&mut self, witness: ChunkStateWitness) -> Result<(), Error> {
+        // TODO(#10265): We'll need to fetch some data from the chain; at the very least we need
+        // the previous block to exist, and we need the previous chunks' receipt roots.
+        // Some of this depends on delayed chunk execution. Also, if the previous block
+        // does not exist, we should queue this (similar to orphans) to retry later.
+        // For now though, we just pass it to the chunk validation logic.
+        self.chunk_validator.start_validating_chunk(witness)
+    }
+
+    /// Distributes the chunk state witness to chunk validators that are
+    /// selected to validate this chunk.
+    pub fn send_chunk_state_witness_to_chunk_validators(
```

Comment:
This is little confusing. Is this client used by both chunk proposer AND chunk validator?

Reply:
Yes, Client is used by everyone so this is just an additional function on the Client. I didn't want to put it in client.rs, so that I don't further increase the size of that file.